Приложение Filmorate разработано согласно ТЗ-11

aigeley opened the pull request.

aigeley committed:
feat: добавить сохранение в БД

aigeley committed:
feat: добавить жанры и рейтинги MPA

aigeley committed:
fix: получать популярные фильмы из БД

aigeley committed:
fix: использовать UTF-8 в ответах GET

aigeley committed:
fix: убрать тест с пустым ID

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,68 @@
+    friend_id      BIGINT,
+    CONSTRAINT user_friends_user_fk FOREIGN KEY (user_id) REFERENCES users (user_id) ON DELETE CASCADE,
+    CONSTRAINT user_friends_friend_fk FOREIGN KEY (friend_id) REFERENCES users (user_id) ON DELETE CASCADE,
+    CONSTRAINT user_friends_uq UNIQUE (user_id, friend_id)
```
Можно было бы использовать просто составной первичный ключ 
PRIMARY KEY (friend_id,user_id)
Чтобы не вводить дополнительное поле)

aigeley replied:
Исправлено
[fix: использовать составной ключ для таблиц-связок](https://github.com/aigeley/java-filmorate/commit/f88d8623f96cece44b40dd93d79df3bc2a099192)

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил мелкие недочеты и то, что можно сделать лучше :) Доброй ночи! Если возникнут вопросы - пиши в слак, буду рад помочь😇

aigeley committed:
fix: использовать составной ключ для таблиц-связок

aigeley committed:
fix: count в getPopularFilms строго больше 0

aigeley committed:
fix: прикреплять рейтинг к фильму в одном запросе

aigeley committed:
fix: добавить LikeStorage и FriendStorage

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -36,7 +36,7 @@ public void deleteLike(@PathVariable("id") long filmId, @PathVariable long userI
 
-    @GetMapping("/popular")
+    @GetMapping(value = "/popular", produces = "application/json;charset=UTF-8")
     public List<Film> getPopularFilms(@RequestParam(defaultValue = DEFAULT_FILMS_TO_DISPLAY) int count) {
```
Исправлено
[fix: count в getPopularFilms строго больше 0](https://github.com/aigeley/java-filmorate/commit/6bc07d6767d00b7fd147ddb96ceb07fb8d38a62e)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,35 @@
+import ru.yandex.practicum.filmorate.storage.DbGenreStorage;
+
+import java.util.Collection;
+
```
Исправлено
[fix: count в getPopularFilms строго больше 0](https://github.com/aigeley/java-filmorate/commit/6bc07d6767d00b7fd147ddb96ceb07fb8d38a62e)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -10,7 +10,9 @@
+import java.util.Objects;
 import java.util.Set;
 
 @With
```
Исправлено
[fix: добавить LikeStorage и FriendStorage](https://github.com/aigeley/java-filmorate/commit/7a1f8019e9296ae2b7079a01971285851cfe6389)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -26,17 +29,44 @@ public class Film implements Identifiable<Film> {
     @Positive
     int duration;
     Set<Long> likes;
```
Исправлено
[fix: добавить LikeStorage и FriendStorage](https://github.com/aigeley/java-filmorate/commit/7a1f8019e9296ae2b7079a01971285851cfe6389)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,243 @@
+import java.util.List;
+import java.util.Set;
+
+@Component
```
Исправлено
[fix: добавить LikeStorage и FriendStorage](https://github.com/aigeley/java-filmorate/commit/7a1f8019e9296ae2b7079a01971285851cfe6389)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbUserStorage.java:
```diff
@@ -0,0 +1,184 @@
+
+    @Override
+    public List<User> getFriends(long userId) {
+        return get(userId)
```
Исправлено
[fix: добавить LikeStorage и FriendStorage](https://github.com/aigeley/java-filmorate/commit/7a1f8019e9296ae2b7079a01971285851cfe6389)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbUserStorage.java:
```diff
@@ -0,0 +1,184 @@
+
+    @Override
+    public List<User> getCommonFriends(long userId, long otherUserId) {
+        return get(userId)
```
Исправлено
[fix: добавить LikeStorage и FriendStorage](https://github.com/aigeley/java-filmorate/commit/7a1f8019e9296ae2b7079a01971285851cfe6389)

aigeley commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,68 @@
+    genre_id      INT,
+    CONSTRAINT film_genres_film_fk FOREIGN KEY (film_id) REFERENCES films (film_id) ON DELETE CASCADE,
+    CONSTRAINT film_genres_genre_fk FOREIGN KEY (genre_id) REFERENCES genres (genre_id) ON DELETE RESTRICT,
+    CONSTRAINT film_genres_uq UNIQUE (film_id, genre_id)
```
Исправлено
[fix: использовать составной ключ для таблиц-связок](https://github.com/aigeley/java-filmorate/commit/f88d8623f96cece44b40dd93d79df3bc2a099192)

aigeley commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,68 @@
+    user_id BIGINT,
+    CONSTRAINT likes_film_fk FOREIGN KEY (film_id) REFERENCES films (film_id) ON DELETE CASCADE,
+    CONSTRAINT likes_user_fk FOREIGN KEY (user_id) REFERENCES users (user_id) ON DELETE CASCADE,
+    CONSTRAINT likes_uq UNIQUE (film_id, user_id)
```
Исправлено
[fix: использовать составной ключ для таблиц-связок](https://github.com/aigeley/java-filmorate/commit/f88d8623f96cece44b40dd93d79df3bc2a099192)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    @Override
+    public Film get(long filmId) {
```
Исправлено
[fix: прикреплять рейтинг к фильму в одном запросе](https://github.com/aigeley/java-filmorate/commit/dc805617d276adda863d51dae6299075c216f883)

aigeley commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,243 @@
+
+    @Override
+    public Collection<Film> getAll() {
+        String sql = "SELECT film_id, film_name, description, release_date, duration, mpa_id FROM films";
```
Исправлено
[fix: прикреплять рейтинг к фильму в одном запросе](https://github.com/aigeley/java-filmorate/commit/dc805617d276adda863d51dae6299075c216f883)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,189 @@
+import java.util.Set;
+
+@Component
+public class DbFilmStorage implements FilmStorage, RowMapper<Film> {
```
Хорошо было бы использовать именование вида ==>>
Интерфейс FilmStorage, класс, его имплементирующий FilmStorageImpl. Так же хорошо было бы создать внутри пакета storage пакет impl, чтобы поместить в него все имплементации)

aigeley replied:
Исправлено
[fix: устранить замечания ревью](https://github.com/aigeley/java-filmorate/pull/3/commits/5f9787407df7d54a5ae200974ace6b3e5e66c868)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,189 @@
+
+    @Override
+    public Film get(long filmId) {
+        String sql = "SELECT f.film_id, f.film_name, f.description, f.release_date, f.duration, f.mpa_id, m.mpa_name " +
```
Можно было бы использовать запрос вида)
select * from FILMS f, MPA m where f.MPA_ID = m.MPA_ID AND FILM_ID = ?
Аналогично с получением всех фильмов)

aigeley replied:
Прошу уточнить, правильно понял, что при простых соединениях двух таблиц, для краткости записи, лучше использовать WHERE вместо JOIN?
Переписал запрос так:
SELECT f.film_id, f.film_name, f.description, f.release_date, f.duration, f.mpa_id, m.mpa_name 
FROM films f, mpa m WHERE m.mpa_id = f.mpa_id AND f.film_id = ?

TyutterinYakov reviewed (approved):
Добрый день! Поздравляю с успешной сдачей спринта! Сил и терпения на пути к цели :) Хорошего дня🌅

aigeley committed:
fix: устранить замечания ревью

- добавить пакет impl
- использовать WHERE вместо JOIN для простых соединений двух таблиц
